### PR TITLE
dbeaver/dbeaver#26613 Fix json binary

### LINKED ISSUE #26613: Binary columns cause JSON export to be invalid
### Description

A table that contains a binary column causes the JSON generated by the "Export data" feature to be invalid.

For example,

```
{
"select * from t1": [
	{
		"id" : 0x1234
	},
	{
		"id" : 0x5678
	}
]}
```

The export menu offers to choose an "Encoding" for "Binaries", but none of the options produces valid JSON:

* Native: `"id" : 0x1234`
* Binary: `"id" : 4` (including a non-printable character before the "4"!)
* Base64: `"id" : EjQ=`
* Hex: `"id" : 0x1234` (same as Native)

### DBeaver Version

Community Edition 24.0.3.202404222221

### Operating System

Linux

### Database and driver

Database: MySQL (Percona) v8.1.0
Driver: MySQL (I don't see any version number)

### Steps to reproduce

Create a simple table, insert some binary data, select it:

```
create table t1 (
	id varbinary(10)
);

insert into t1 values (0x1234),(0x5678);

select * from t1;
```

Click "Export data", choose "JSON / Export to JSON file(s)"

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by HocKu7**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/stream/StreamTransferConsumer.java [hunk 1]
```diff
@@ -1065,14 +1066,22 @@ public void writeBinaryData(@NotNull DBDContentStorage cs) throws IOException {
...
-                            // Binary stream
-                            try (Reader reader = new InputStreamReader(stream, cs.getCharset())) {
-                                IOUtils.copyText(reader, writer);
+                            byte[] readBuffer = new byte[1000];
+                            for (; ; ) {
+                                int count = stream.read(readBuffer);
+                                if (count <= 0) {
+                                    break;
+                                }
+                                String content = new String(readBuffer, 0, count, settings.getOutputEncoding());
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/stream/StreamTransferConsumer.java]: You should be using `cs.getCharset()` instead.

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/stream/StreamTransferConsumer.java [hunk 2]
```diff
@@ -1065,14 +1066,22 @@ public void writeBinaryData(@NotNull DBDContentStorage cs) throws IOException {
                                 writer.write(binaryString);
                                 break;
                             }
+                            break;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/stream/StreamTransferConsumer.java:1073]: Remove the `break` above then

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/data/json/JSONUtils.java [hunk 3]
```diff
@@ -104,7 +104,13 @@ public static String escapeJsonString(String str) {
                 case '\f' -> result.append("\\f");
                 case '\b' -> result.append("\\b");
                 case '"', '\\', '/' -> result.append("\\").append(c);
-                default -> result.append(c);
+                default -> {
+                    if ((int) c < 32) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/data/json/JSONUtils.java:108]:
```suggestion
                    if (c < ' ') {
                        // escape non-displayable ASCII characters
```

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/stream/StreamTransferConsumer.java (hunk 1 above)
Comment (HocKu7) [reply plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/stream/StreamTransferConsumer.java]: I guess it was a mistake to use `cs.getCharset()` before. As I see there is need to take charset from params, cause we have such properties in configuration page

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/data/json/JSONUtils.java (hunk 3 above)
Comment (HocKu7) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/data/json/JSONUtils.java:108]: It has not so clear intention. I prefer to leave it

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/stream/StreamTransferConsumer.java (hunk 1 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/stream/StreamTransferConsumer.java]: But this is not correct. You are reading data from `cs` and should use its charset, not the output charset. The output charset is already used by the underlying stream.

**Review by ShadelessFox — APPROVED** [review]

**Comment by MashaKorax** [comment]
verified

**Pull request merged**